## Setup Ollama :

In [1]:
!sudo apt update
!sudo apt install -y pciutils
!curl -fsSL https://ollama.com/install.sh | sh

import threading
import subprocess
import time

def run_ollama_serve():
  subprocess.Popen(["ollama", "serve"])

thread = threading.Thread(target=run_ollama_serve)
thread.start()
time.sleep(5)

!ollama pull llama3.1:8b
!ollama list

Get:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,626 B]
Hit:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease
Get:3 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]
Hit:4 http://archive.ubuntu.com/ubuntu jammy InRelease
Get:5 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]
Hit:6 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease
Get:7 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]
Hit:8 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease
Get:9 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages [2,469 kB]
Hit:10 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease
Get:11 https://r2u.stat.illinois.edu/ubuntu jammy/main amd64 Packages [2,629 kB]
Get:12 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]
Get:13 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packa

## Clone the Invoices:

In [2]:
# Download ZIP of the whole repository
!wget https://github.com/MasrourTawfik/Textra_Insights/archive/refs/heads/main.zip

# Unzip the folder
!unzip main.zip


--2024-12-15 15:15:21--  https://github.com/MasrourTawfik/Textra_Insights/archive/refs/heads/main.zip
Resolving github.com (github.com)... 20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://codeload.github.com/MasrourTawfik/Textra_Insights/zip/refs/heads/main [following]
--2024-12-15 15:15:21--  https://codeload.github.com/MasrourTawfik/Textra_Insights/zip/refs/heads/main
Resolving codeload.github.com (codeload.github.com)... 20.205.243.165
Connecting to codeload.github.com (codeload.github.com)|20.205.243.165|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘main.zip’

main.zip                [               <=>  ]  60.77M  13.5MB/s    in 4.4s    

2024-12-15 15:15:26 (13.7 MB/s) - ‘main.zip’ saved [63720918]

Archive:  main.zip
ea279ff8180ee41cb62c54b29e32d559a92c426c
   creating: Textra_Insights-main/
  inflating: T

In [3]:
import os

Invoices_path = "/content/Textra_Insights-main/Files/Invoices"
print("Number of Images : ",len(os.listdir(Invoices_path)))

Number of Images :  134


- We'll not work with 134 invoices !!, then it will be hard to manually find the appropriate account.
So we can select just 34 Invoices with diverse accounts.

In [4]:
import os
import shutil

New_Folder_Path = "/content/Selected_Invoices"
Selected_Invoices = ['3.jpg', '49.jpg', '57.jpg', '58.jpg', '61.jpg', '66.jpg', '67.jpg', '76.jpg',
                     '109.jpg', '114.jpg', '121.jpg', '129.jpg', '131.jpg', '151.jpg', '164.jpg',
                     '311.jpg', '314.jpg', '315.jpg', '316.jpg', '317.jpg', '319.jpg', '327.jpg',
                     '340.jpg', '383.jpg', '417.jpg', '419.jpg', '45.jpg', '9.jpg', '52.jpg', '80.jpg',
                     '4.jpg', '27.jpg', '351.jpg', '170.jpg']

os.makedirs(New_Folder_Path, exist_ok=True)

for file_name in os.listdir(Invoices_path):
    if file_name in Selected_Invoices:
        shutil.copy(os.path.join(Invoices_path, file_name), New_Folder_Path)

print(f"Selected images have been copied to {New_Folder_Path}")


Selected images have been copied to /content/Selected_Invoices


## OCR

### Setup Paddle

In [5]:
## CPU :
#!pip install paddlepaddle -i https://pypi.tuna.tsinghua.edu.cn/simple
## GPU :
!pip install paddlepaddle-gpu -i https://pypi.tuna.tsinghua.edu.cn/simple

Looking in indexes: https://pypi.tuna.tsinghua.edu.cn/simple
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 758.9/758.9 MB 1.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.5/65.5 kB 5.8 MB/s eta 0:00:00
  Attempting uninstall: opt-einsum
    Found existing installation: opt_einsum 3.4.0
    Uninstalling opt_einsum-3.4.0:
      Successfully uninstalled opt_einsum-3.4.0


In [6]:
!pip install "paddleocr>=2.0.1" # Recommend to use version 2.0.1+

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 87.2/87.2 kB 7.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 544.7/544.7 kB 34.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 161.9/161.9 kB 14.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 294.9/294.9 kB 24.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 912.2/912.2 kB 54.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 244.3/244.3 kB 20.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 81.7 MB/s eta 0:00:00
  Created wheel for fire: filename=fire-0.7.0-py3-none-any.whl size=114249 sha256=de4c119ac6f5422bac6e97bf1ac7f038f26040c0d53ac24be8d94795321ecf07
  Stored in directory: /root/.cache/pip/wheels/19/39/2f/2d3cadc408a8804103f1c34ddd4b9f6a93497b11fa96fe738e
Successfully built fire
  Attempting uninstall: albucore
    Found existing installation: albucore 0.0.19
    Uninstalling albucore-0.0.19:
   

### Run On Selected Invoices

In [7]:
import time
import os
from tqdm import tqdm
import pandas as pd

################################################################################
################################### PaddleOCR ##################################
from paddleocr import PaddleOCR
paddle_reader = PaddleOCR(use_angle_cls=True, lang='fr')

def Run_Paddle(Image_path):
    Start = time.time()
    result = paddle_reader.ocr(Image_path, det=True, rec=True)
    End = time.time()
    # Inference Time
    Time = End-Start
    # Process the result
    extracted_text = ''
    #print("Lenght Result: ",len(result))
    for line in result:
        if line is not None:
            for word in line:
                #print("Word: ",word)
                # Access the text part of the tuple
                extracted_text += word[1][0] + ' '  # Access the first element of the recognized text (the text itself)
            extracted_text += '\n'
    return extracted_text,Time


def Run_Paddle_ALL(Folder_path):
    df = pd.DataFrame(columns=['InvoiceName', 'Text', 'Time(s)'])
    Files = os.listdir(Folder_path)
    Files = sorted(Files, key=lambda x: int(x.split('.')[0]))
    Texts = []
    Times = []
    for file in tqdm(Files,desc="Paddle Processing", unit="number"):
        print(f"{file} is processing")
        Invoice_path = os.path.join(Folder_path, file)
        Text_paddle,Time_paddle = Run_Paddle(Invoice_path)
        Texts.append(Text_paddle)
        Times.append(Time_paddle)
    df['InvoiceName'] = Files
    df['Text'] = Texts
    df['Time(s)'] = Times

    # Saving the DataFrame to a CSV file
    df.to_csv('PaddleOCR_Results.csv', index=False)

download https://paddleocr.bj.bcebos.com/PP-OCRv3/english/en_PP-OCRv3_det_infer.tar to /root/.paddleocr/whl/det/en/en_PP-OCRv3_det_infer/en_PP-OCRv3_det_infer.tar


100%|██████████| 3910/3910 [00:00<00:00, 4366.10it/s]


download https://paddleocr.bj.bcebos.com/PP-OCRv3/multilingual/latin_PP-OCRv3_rec_infer.tar to /root/.paddleocr/whl/rec/latin/latin_PP-OCRv3_rec_infer/latin_PP-OCRv3_rec_infer.tar


100%|██████████| 9930/9930 [00:01<00:00, 9286.46it/s] 


download https://paddleocr.bj.bcebos.com/dygraph_v2.0/ch/ch_ppocr_mobile_v2.0_cls_infer.tar to /root/.paddleocr/whl/cls/ch_ppocr_mobile_v2.0_cls_infer/ch_ppocr_mobile_v2.0_cls_infer.tar


100%|██████████| 2138/2138 [00:00<00:00, 2911.65it/s]

[2024/12/15 15:17:12] ppocr DEBUG: Namespace(help='==SUPPRESS==', use_gpu=True, use_xpu=False, use_npu=False, use_mlu=False, ir_optim=True, use_tensorrt=False, min_subgraph_size=15, precision='fp32', gpu_mem=500, gpu_id=0, image_dir=None, page_num=0, det_algorithm='DB', det_model_dir='/root/.paddleocr/whl/det/en/en_PP-OCRv3_det_infer', det_limit_side_len=960, det_limit_type='max', det_box_type='quad', det_db_thresh=0.3, det_db_box_thresh=0.6, det_db_unclip_ratio=1.5, max_batch_size=10, use_dilation=False, det_db_score_mode='fast', det_east_score_thresh=0.8, det_east_cover_thresh=0.1, det_east_nms_thresh=0.2, det_sast_score_thresh=0.5, det_sast_nms_thresh=0.2, det_pse_thresh=0, det_pse_box_thresh=0.85, det_pse_min_area=16, det_pse_scale=1, scales=[8, 16, 32], alpha=1.0, beta=1.0, fourier_degree=5, rec_algorithm='SVTR_LCNet', rec_model_dir='/root/.paddleocr/whl/rec/latin/latin_PP-OCRv3_rec_infer', rec_image_inverse=True, rec_image_shape='3, 48, 320', rec_batch_num=6, max_text_length=25, 

[2024/12/15 15:17:13] ppocr WARNING: The first GPU is used for inference by default, GPU ID: 0
[2024/12/15 15:17:13] ppocr WARNING: The first GPU is used for inference by default, GPU ID: 0


In [8]:
Run_Paddle_ALL(New_Folder_Path)

Paddle Processing:   0%|          | 0/34 [00:00<?, ?number/s]

3.jpg is processing
[2024/12/15 15:17:16] ppocr DEBUG: dt_boxes num : 36, elapsed : 1.858729600906372
[2024/12/15 15:17:16] ppocr DEBUG: cls num  : 36, elapsed : 0.19477224349975586
[2024/12/15 15:17:16] ppocr DEBUG: rec_res num  : 36, elapsed : 0.20941948890686035


Paddle Processing:   3%|▎         | 1/34 [00:02<01:18,  2.39s/number]

4.jpg is processing
[2024/12/15 15:17:17] ppocr DEBUG: dt_boxes num : 46, elapsed : 0.08150887489318848
[2024/12/15 15:17:17] ppocr DEBUG: cls num  : 46, elapsed : 0.12769770622253418
[2024/12/15 15:17:17] ppocr DEBUG: rec_res num  : 46, elapsed : 0.16458773612976074


Paddle Processing:   6%|▌         | 2/34 [00:02<00:40,  1.25s/number]

9.jpg is processing
[2024/12/15 15:17:17] ppocr DEBUG: dt_boxes num : 43, elapsed : 0.10197257995605469
[2024/12/15 15:17:17] ppocr DEBUG: cls num  : 43, elapsed : 0.10600733757019043
[2024/12/15 15:17:17] ppocr DEBUG: rec_res num  : 43, elapsed : 0.1934499740600586


Paddle Processing:   9%|▉         | 3/34 [00:03<00:28,  1.10number/s]

27.jpg is processing
[2024/12/15 15:17:18] ppocr DEBUG: dt_boxes num : 58, elapsed : 0.09360623359680176
[2024/12/15 15:17:18] ppocr DEBUG: cls num  : 58, elapsed : 0.10368847846984863
[2024/12/15 15:17:18] ppocr DEBUG: rec_res num  : 58, elapsed : 0.15052318572998047


Paddle Processing:  12%|█▏        | 4/34 [00:03<00:21,  1.39number/s]

45.jpg is processing
[2024/12/15 15:17:18] ppocr DEBUG: dt_boxes num : 36, elapsed : 0.07700562477111816
[2024/12/15 15:17:18] ppocr DEBUG: cls num  : 36, elapsed : 0.08145284652709961
[2024/12/15 15:17:18] ppocr DEBUG: rec_res num  : 36, elapsed : 0.12151288986206055


Paddle Processing:  15%|█▍        | 5/34 [00:04<00:17,  1.69number/s]

49.jpg is processing
[2024/12/15 15:17:18] ppocr DEBUG: dt_boxes num : 98, elapsed : 0.09698724746704102
[2024/12/15 15:17:19] ppocr DEBUG: cls num  : 98, elapsed : 0.18875813484191895
[2024/12/15 15:17:19] ppocr DEBUG: rec_res num  : 98, elapsed : 0.28429627418518066


Paddle Processing:  18%|█▊        | 6/34 [00:04<00:17,  1.62number/s]

52.jpg is processing
[2024/12/15 15:17:19] ppocr DEBUG: dt_boxes num : 63, elapsed : 0.06802558898925781
[2024/12/15 15:17:19] ppocr DEBUG: cls num  : 63, elapsed : 0.10064411163330078
[2024/12/15 15:17:19] ppocr DEBUG: rec_res num  : 63, elapsed : 0.12488389015197754


Paddle Processing:  21%|██        | 7/34 [00:05<00:14,  1.88number/s]

57.jpg is processing
[2024/12/15 15:17:19] ppocr DEBUG: dt_boxes num : 34, elapsed : 0.04385566711425781
[2024/12/15 15:17:19] ppocr DEBUG: cls num  : 34, elapsed : 0.048002004623413086
[2024/12/15 15:17:19] ppocr DEBUG: rec_res num  : 34, elapsed : 0.09823822975158691


Paddle Processing:  24%|██▎       | 8/34 [00:05<00:11,  2.29number/s]

58.jpg is processing
[2024/12/15 15:17:20] ppocr DEBUG: dt_boxes num : 61, elapsed : 0.05466604232788086
[2024/12/15 15:17:20] ppocr DEBUG: cls num  : 61, elapsed : 0.08050775527954102
[2024/12/15 15:17:20] ppocr DEBUG: rec_res num  : 61, elapsed : 0.1493697166442871


Paddle Processing:  26%|██▋       | 9/34 [00:05<00:10,  2.46number/s]

61.jpg is processing
[2024/12/15 15:17:20] ppocr DEBUG: dt_boxes num : 44, elapsed : 0.047135353088378906
[2024/12/15 15:17:20] ppocr DEBUG: cls num  : 44, elapsed : 0.08344244956970215
[2024/12/15 15:17:20] ppocr DEBUG: rec_res num  : 44, elapsed : 0.10750985145568848


Paddle Processing:  29%|██▉       | 10/34 [00:06<00:08,  2.70number/s]

66.jpg is processing
[2024/12/15 15:17:20] ppocr DEBUG: dt_boxes num : 98, elapsed : 0.06918549537658691
[2024/12/15 15:17:20] ppocr DEBUG: cls num  : 98, elapsed : 0.12991929054260254
[2024/12/15 15:17:21] ppocr DEBUG: rec_res num  : 98, elapsed : 0.16504573822021484


Paddle Processing:  32%|███▏      | 11/34 [00:06<00:08,  2.58number/s]

67.jpg is processing
[2024/12/15 15:17:21] ppocr DEBUG: dt_boxes num : 105, elapsed : 0.06946516036987305
[2024/12/15 15:17:21] ppocr DEBUG: cls num  : 105, elapsed : 0.14185738563537598
[2024/12/15 15:17:21] ppocr DEBUG: rec_res num  : 105, elapsed : 0.2017979621887207


Paddle Processing:  35%|███▌      | 12/34 [00:06<00:09,  2.41number/s]

76.jpg is processing
[2024/12/15 15:17:21] ppocr DEBUG: dt_boxes num : 51, elapsed : 0.059636592864990234
[2024/12/15 15:17:21] ppocr DEBUG: cls num  : 51, elapsed : 0.07122206687927246
[2024/12/15 15:17:21] ppocr DEBUG: rec_res num  : 51, elapsed : 0.1197967529296875


Paddle Processing:  38%|███▊      | 13/34 [00:07<00:08,  2.62number/s]

80.jpg is processing
[2024/12/15 15:17:21] ppocr DEBUG: dt_boxes num : 29, elapsed : 0.045024871826171875
[2024/12/15 15:17:21] ppocr DEBUG: cls num  : 29, elapsed : 0.04593086242675781
[2024/12/15 15:17:22] ppocr DEBUG: rec_res num  : 29, elapsed : 0.09401822090148926


Paddle Processing:  41%|████      | 14/34 [00:07<00:06,  2.99number/s]

109.jpg is processing
[2024/12/15 15:17:22] ppocr DEBUG: dt_boxes num : 39, elapsed : 0.0465397834777832
[2024/12/15 15:17:22] ppocr DEBUG: cls num  : 39, elapsed : 0.049924612045288086
[2024/12/15 15:17:22] ppocr DEBUG: rec_res num  : 39, elapsed : 0.09537339210510254


Paddle Processing:  44%|████▍     | 15/34 [00:07<00:05,  3.24number/s]

114.jpg is processing
[2024/12/15 15:17:22] ppocr DEBUG: dt_boxes num : 90, elapsed : 0.06253385543823242
[2024/12/15 15:17:22] ppocr DEBUG: cls num  : 90, elapsed : 0.13147687911987305
[2024/12/15 15:17:22] ppocr DEBUG: rec_res num  : 90, elapsed : 0.2173480987548828


Paddle Processing:  47%|████▋     | 16/34 [00:08<00:06,  2.80number/s]

121.jpg is processing
[2024/12/15 15:17:22] ppocr DEBUG: dt_boxes num : 34, elapsed : 0.04126763343811035
[2024/12/15 15:17:22] ppocr DEBUG: cls num  : 34, elapsed : 0.04823923110961914
[2024/12/15 15:17:22] ppocr DEBUG: rec_res num  : 34, elapsed : 0.08177709579467773


Paddle Processing:  50%|█████     | 17/34 [00:08<00:05,  3.19number/s]

129.jpg is processing
[2024/12/15 15:17:23] ppocr DEBUG: dt_boxes num : 96, elapsed : 0.06784772872924805
[2024/12/15 15:17:23] ppocr DEBUG: cls num  : 96, elapsed : 0.11960148811340332
[2024/12/15 15:17:23] ppocr DEBUG: rec_res num  : 96, elapsed : 0.19312191009521484


Paddle Processing:  53%|█████▎    | 18/34 [00:08<00:05,  2.83number/s]

131.jpg is processing
[2024/12/15 15:17:23] ppocr DEBUG: dt_boxes num : 28, elapsed : 0.04982113838195801
[2024/12/15 15:17:23] ppocr DEBUG: cls num  : 28, elapsed : 0.05020594596862793
[2024/12/15 15:17:23] ppocr DEBUG: rec_res num  : 28, elapsed : 0.07469034194946289


Paddle Processing:  56%|█████▌    | 19/34 [00:09<00:04,  3.12number/s]

151.jpg is processing
[2024/12/15 15:17:23] ppocr DEBUG: dt_boxes num : 74, elapsed : 0.05709481239318848
[2024/12/15 15:17:23] ppocr DEBUG: cls num  : 74, elapsed : 0.08773517608642578
[2024/12/15 15:17:24] ppocr DEBUG: rec_res num  : 74, elapsed : 0.13400888442993164


Paddle Processing:  59%|█████▉    | 20/34 [00:09<00:04,  3.05number/s]

164.jpg is processing
[2024/12/15 15:17:24] ppocr DEBUG: dt_boxes num : 58, elapsed : 0.053904056549072266
[2024/12/15 15:17:24] ppocr DEBUG: cls num  : 58, elapsed : 0.07116341590881348
[2024/12/15 15:17:24] ppocr DEBUG: rec_res num  : 58, elapsed : 0.1395277976989746


Paddle Processing:  62%|██████▏   | 21/34 [00:09<00:04,  3.08number/s]

170.jpg is processing
[2024/12/15 15:17:24] ppocr DEBUG: dt_boxes num : 38, elapsed : 0.04883241653442383
[2024/12/15 15:17:24] ppocr DEBUG: cls num  : 38, elapsed : 0.06081271171569824
[2024/12/15 15:17:24] ppocr DEBUG: rec_res num  : 38, elapsed : 0.1688063144683838


Paddle Processing:  65%|██████▍   | 22/34 [00:10<00:03,  3.08number/s]

311.jpg is processing
[2024/12/15 15:17:24] ppocr DEBUG: dt_boxes num : 41, elapsed : 0.05649876594543457
[2024/12/15 15:17:24] ppocr DEBUG: cls num  : 41, elapsed : 0.051122426986694336
[2024/12/15 15:17:24] ppocr DEBUG: rec_res num  : 41, elapsed : 0.08512401580810547


Paddle Processing:  68%|██████▊   | 23/34 [00:10<00:03,  3.35number/s]

314.jpg is processing
[2024/12/15 15:17:24] ppocr DEBUG: dt_boxes num : 41, elapsed : 0.05527782440185547
[2024/12/15 15:17:25] ppocr DEBUG: cls num  : 41, elapsed : 0.050672292709350586
[2024/12/15 15:17:25] ppocr DEBUG: rec_res num  : 41, elapsed : 0.10654067993164062


Paddle Processing:  71%|███████   | 24/34 [00:10<00:02,  3.49number/s]

315.jpg is processing
[2024/12/15 15:17:25] ppocr DEBUG: dt_boxes num : 124, elapsed : 0.08230376243591309
[2024/12/15 15:17:25] ppocr DEBUG: cls num  : 124, elapsed : 0.15485382080078125
[2024/12/15 15:17:25] ppocr DEBUG: rec_res num  : 124, elapsed : 0.27672815322875977


Paddle Processing:  74%|███████▎  | 25/34 [00:11<00:03,  2.63number/s]

316.jpg is processing
[2024/12/15 15:17:25] ppocr DEBUG: dt_boxes num : 90, elapsed : 0.06943368911743164
[2024/12/15 15:17:25] ppocr DEBUG: cls num  : 90, elapsed : 0.10424518585205078
[2024/12/15 15:17:26] ppocr DEBUG: rec_res num  : 90, elapsed : 0.178924560546875


Paddle Processing:  76%|███████▋  | 26/34 [00:11<00:03,  2.58number/s]

317.jpg is processing
[2024/12/15 15:17:26] ppocr DEBUG: dt_boxes num : 81, elapsed : 0.06430840492248535
[2024/12/15 15:17:26] ppocr DEBUG: cls num  : 81, elapsed : 0.10048317909240723
[2024/12/15 15:17:26] ppocr DEBUG: rec_res num  : 81, elapsed : 0.17710351943969727


Paddle Processing:  79%|███████▉  | 27/34 [00:11<00:02,  2.54number/s]

319.jpg is processing
[2024/12/15 15:17:26] ppocr DEBUG: dt_boxes num : 55, elapsed : 0.05774569511413574
[2024/12/15 15:17:26] ppocr DEBUG: cls num  : 55, elapsed : 0.08767223358154297
[2024/12/15 15:17:26] ppocr DEBUG: rec_res num  : 55, elapsed : 0.13567304611206055


Paddle Processing:  82%|████████▏ | 28/34 [00:12<00:02,  2.65number/s]

327.jpg is processing
[2024/12/15 15:17:27] ppocr DEBUG: dt_boxes num : 68, elapsed : 0.07443618774414062
[2024/12/15 15:17:27] ppocr DEBUG: cls num  : 68, elapsed : 0.08159637451171875
[2024/12/15 15:17:27] ppocr DEBUG: rec_res num  : 68, elapsed : 0.14477777481079102


Paddle Processing:  85%|████████▌ | 29/34 [00:12<00:01,  2.68number/s]

340.jpg is processing
[2024/12/15 15:17:27] ppocr DEBUG: dt_boxes num : 48, elapsed : 0.057707786560058594
[2024/12/15 15:17:27] ppocr DEBUG: cls num  : 48, elapsed : 0.05661201477050781
[2024/12/15 15:17:27] ppocr DEBUG: rec_res num  : 48, elapsed : 0.1314866542816162


Paddle Processing:  88%|████████▊ | 30/34 [00:12<00:01,  2.82number/s]

351.jpg is processing
[2024/12/15 15:17:27] ppocr DEBUG: dt_boxes num : 58, elapsed : 0.06053876876831055
[2024/12/15 15:17:27] ppocr DEBUG: cls num  : 58, elapsed : 0.08646821975708008
[2024/12/15 15:17:27] ppocr DEBUG: rec_res num  : 58, elapsed : 0.17112374305725098


Paddle Processing:  91%|█████████ | 31/34 [00:13<00:01,  2.72number/s]

383.jpg is processing
[2024/12/15 15:17:28] ppocr DEBUG: dt_boxes num : 45, elapsed : 0.04772663116455078
[2024/12/15 15:17:28] ppocr DEBUG: cls num  : 45, elapsed : 0.05616021156311035
[2024/12/15 15:17:28] ppocr DEBUG: rec_res num  : 45, elapsed : 0.0926971435546875


Paddle Processing:  94%|█████████▍| 32/34 [00:13<00:00,  3.01number/s]

417.jpg is processing
[2024/12/15 15:17:28] ppocr DEBUG: dt_boxes num : 95, elapsed : 0.08538532257080078
[2024/12/15 15:17:28] ppocr DEBUG: cls num  : 95, elapsed : 0.13228917121887207
[2024/12/15 15:17:28] ppocr DEBUG: rec_res num  : 95, elapsed : 0.1743764877319336


Paddle Processing:  97%|█████████▋| 33/34 [00:14<00:00,  2.79number/s]

419.jpg is processing
[2024/12/15 15:17:28] ppocr DEBUG: dt_boxes num : 41, elapsed : 0.05589938163757324
[2024/12/15 15:17:28] ppocr DEBUG: cls num  : 41, elapsed : 0.0484461784362793
[2024/12/15 15:17:28] ppocr DEBUG: rec_res num  : 41, elapsed : 0.08387899398803711


Paddle Processing: 100%|██████████| 34/34 [00:14<00:00,  2.36number/s]


## LLM

### Setup

In [9]:
!pip install langchain-openai==0.2.9
!pip install openai==1.55.3 httpx==0.27.2 --force-reinstall --quiet

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.4/50.4 kB 4.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 59.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 172.0/172.0 kB 14.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.7/57.7 kB 3.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 389.6/389.6 kB 22.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.4/76.4 kB 6.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.6/78.6 kB 6.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 93.1/93.1 kB 5.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 70.4/70.4 kB 6.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 345.0/345.0 kB 25.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 457.0/457.0 kB 35.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 79.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.5/7

In [ ]:
import os
os.kill(os.getpid(), 9)

### Implimentation

In [1]:
!ollama list

NAME           ID              SIZE      MODIFIED      
llama3.1:8b    46e0c10c039e    4.9 GB    4 minutes ago    


In [2]:
from langchain_core.prompts import ChatPromptTemplate
from pydantic import BaseModel, Field
from langchain_openai import ChatOpenAI

class Description(BaseModel):
    analyse: str = Field("", description="Analyse détaillée de la transaction.")
    proposition: str = Field("", description="Intitulé du compte à débiter dans cette facture.")
    nature : str = Field("", description="Nature du produit ou service acheté (ex. : 'Matériel informatique', 'Fournitures de bureau', etc.).")
    Keywords : str = Field("", description="Les mots-clés dans la facture liés à une opération d'achat. Par exemple : Ordinateur, Service, Machine, Eau, Énergie...")


system = f"""

*Vous êtes un expert comptable chargé d'analyser une facture d'achat en vous basant sur le texte extrait de cette facture. Votre tâche consiste à extraire et à classifier les informations suivantes :*

1. **Analyse détaillée de la transaction** :

    - Fournissez une analyse comptable détaillée de la transaction.

2. **Proposition du compte à débiter** :

   - Sur la base de l'analyse de la transaction, proposez un **compte à débiter** approprié pour cette facture (par exemple, **Fournitures de bureau** ou **Achats de combustibles**).

3. **Nature du produit ou service acheté** :
   - Précisez la **nature spécifique** du produit ou service acheté, comme par exemple **Matériel informatique**, **Fournitures de bureau**, **Prestations de services**, etc.

4. **Extraction des mots-clés** :
  - Identifiez les mots-clés directement liés à l'achat mentionnés dans la facture. Ces mots-clés doivent correspondre aux éléments caractéristiques de l'achat (exemple : Ordinateur, Service, Machine, Eau, Énergie, etc.). Assurez-vous que ces mots-clés sont représentatifs et facilitent la classification des informations.


### **Exemple d'entrée** :

```
Achat de 10 ordinateurs portables et 2 imprimantes pour l'entreprise.

```

### **Exemple de sortie attendue** :

```
analyse: "L'achat concerne du matériel informatique destiné à être utilisé de manière durable dans les opérations de l'entreprise. Il s'agit donc d'une immobilisation."
proposition: "215 - Matériel informatique"
nature: "Matériel informatique"
Keywords: "Ordinateur, Imprimante"
```
"""

## Prompt Template
prompt = ChatPromptTemplate.from_messages(
    [
        ("system", system),
        ("human", """
        Texte extrait à partir de la facture ici :\n {OCR_Output} \n
        Quelle est votre analyse de cette facture ?
        """),
    ]
)

LLM = ChatOpenAI(model="llama3.1:8b",
    #temperature=0,
    api_key='ollama',
    base_url="http://localhost:11434/v1",
)
LLM_Structered = LLM.with_structured_output(Description)
LLM_descriptor = prompt | LLM_Structered

In [3]:
def Run_LLM(OCR_Output):
    Input = {"OCR_Output":OCR_Output}
    response = LLM_descriptor.invoke(Input)
    return response


In [4]:
import pandas as pd

OCR_Results = pd.read_csv('PaddleOCR_Results.csv')
Invoice_Text = OCR_Results['Text'][5]
Invoice_Text

'Référence :  0026757005: Votre Direction Préfectorale CASA-ANFA Facture Eau N° 1507042573AM p xbillg cl 8gL 48, Rue Mohamed Diouri Date 30/03/2023 CASABLANCA contact@lydec.co.ma ICE: 001685924000018 www.lydec.ma Lydec lw Centre de relation clientéle 0522312020 P 202304 SMART CALL DISTRIBUTION 3 RUE ABDELKADER ELMAZINI ETG 1 APPT GH CASABLANCA Avis de facturation Eau et slal 3 y g&L yle&) assainissement 18/02/2023 au 22/03/2023 C9 1m9KF 1M3 Période de consommation Eau du Evolution de votre consommation pyal4 lll Consommation Montants en Dhs Eau et assainissement 14 m3 209,59 Jx aifl g s atl Taxes 14,69 Avr Mai Jui Jul AoaSepOct NovDécJan Fév Mar Avr Sous total TTC 224,28 22222222222222222223232323 Votre contrat Net ä payer 224,28 1171819 Mode de réglement Prélévement bancaire Prochain passage de votre encaisseur Votre option tarifaire Montant des impayés TTC ** 0,00 1P? o 1p6r19 q-Activité commerciale et pro  Sauf erreur ou omission 8ui gf Uas Isc Date limite de paiement 18/04/2023 s1s

In [5]:
response= Run_LLM(Invoice_Text)
print("Analyse        :",response.analyse)
print("Proposition    :",response.proposition)
print("Nature         :",response.nature)
print("Keywords       :",response.Keywords)

Analyse        : Cette facture correspond à des consommations d'eau et d'assainissement pour la période du 18/02/2023 au 22/03/2023. Le montant total TTC est de 224,28 Dhs.
Proposition    : Impôts sur les produits
Nature         : Fournitures de services - Eau et assainissement
Keywords       : Eau et assainissement, Taxes professionnelles


In [6]:
Query = response.analyse + "\n" + response.proposition + "\n" + response.nature + "\n" + response.Keywords
print(Query)

Cette facture correspond à des consommations d'eau et d'assainissement pour la période du 18/02/2023 au 22/03/2023. Le montant total TTC est de 224,28 Dhs.
Impôts sur les produits
Fournitures de services - Eau et assainissement
Eau et assainissement, Taxes professionnelles


- You can do the same for the rest of 34 Invoices, and use GPT-4 to give you an approximation of the right account , then  use this approximation to find the exact account with the PCG file manually.

## Why Raw Definitions Are Bad ?

In [7]:
Definition1 = """
6125. Achats non stockés de matières et de fournitures :

Ce compte enregistre les achats de matières et fournitures non stockables tels que l’eau,
l’électricité et autres énergies, ainsi que les achats de matières et fournitures non stockées par
l’entreprise (les matières et fournitures dont la gestion des stocks est jugée inutile par l’entreprise).
"""

Definition2 = """
6126. Achats de travaux, études et prestations de services :

Ce compte enregistre les divers travaux et services d'exploitation confiés, en partie, par l'entreprise à autre personne.
"""

Definition3 = """
6145. Frais postaux et frais de télécommunication :

Ce compte enregistre les frais postaux (achats de timbres, achats d’enveloppes préaffranchies ...),
les frais de téléphone, les frais de transmission de messages dactylographiques (service de télex) et
les frais de télégramme (message transmis par voie électrique ou radioélectrique). Cependant,
les frais engagés pour l’envoi des fax sont aussi à comptabiliser dans ce compte puisqu’ils
font aussi partie des frais de télécommunication.
"""

In [ ]:
!pip install sentence-transformers

In [9]:
from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity

# Charger un modèle pré-entraîné de Sentence Transformers
model = SentenceTransformer('xlm-r-base-en-ko-nli-ststb')

/usr/local/lib/python3.10/dist-packages/sentence_transformers/cross_encoder/CrossEncoder.py:13: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)
  from tqdm.autonotebook import tqdm, trange
/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/229 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/4.00k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/731 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.11G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/500 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/9.10M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/150 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [10]:
# Liste des définitions
definitions = [
    Definition1,
    Definition2,
    Definition3
]


# Encoder la chaîne de réponse une seule fois
embeddings2 = model.encode([Query])

# Itérer sur les définitions et calculer les similarités
for i, definition in enumerate(definitions):
    embeddings1 = model.encode([definition])  # Encoder chaque définition
    similarity = cosine_similarity(embeddings1, embeddings2)  # Calculer la similarité
    print(f"Score de similarité pour la définition {i + 1} : {similarity[0][0]}")


Score de similarité pour la définition 1 : 0.5218554735183716
Score de similarité pour la définition 2 : 0.36586281657218933
Score de similarité pour la définition 3 : 0.41204366087913513
